nnvm/python/nnvm/testing/yolo_detection.py

```diff
@@ -65,72 +65,105 @@ def _box_union(a, b):
+    return
 
+def _get_region_detections(l, im_shape, net_shape, thresh, relative, dets):
+    data = l['output']
+    before_correct_dets = []
+    for row in range(data.shape[2]):
+        for col in range(data.shape[3]):
+            for n in range(data.shape[0]):
+                prob = [0]*l['classes']
+                scale = data[n, l['coords'], row, col] if not l['background'] else 1
+                location = [n, row, col]
+                box_b = _get_box(data, l['biases'], n, location,
+                                 data.shape[2], data.shape[3], data.shape[2], data.shape[3])
+                objectness = scale if scale > thresh else 0
+                if objectness:
+                    prob = scale * data[n, l['coords']+1: l['coords']+1+l['classes'],
+                                        row, col]
+                    prob[prob < thresh] = 0
+                detection = {}
+                detection['bbox'] = box_b
+                detection['prob'] = prob
+                detection['objectness'] = objectness
+                before_correct_dets.append(detection)
+    _correct_boxes(before_correct_dets, im_shape[0], im_shape[1],
+                   net_shape[0], net_shape[1], relative)
+    dets.extend(before_correct_dets)
+    return
+
+def fill_network_boxes(net_shape, im_shape,
+                       thresh, hier_thresh, relative, tvm_out):
+    dets = []
+    for layer in tvm_out:
+        if layer['type'] == 'Yolo':
+            _get_yolo_detections(layer, im_shape, net_shape, thresh, relative, dets)
+        elif layer['type'] == 'Region':
+            _get_region_detections(layer, im_shape, net_shape, thresh, relative, dets)
+    return dets
+
+def do_nms_sort(dets, classes, thresh):
+    "Does the sorting based on the threshold values"
+    k = len(dets)-1
+    cnt = 0
+    while cnt < k:
+        if dets[cnt]['objectness'] == 0:
+            temp = dets[cnt]
+            dets[cnt] = dets[k]
+            dets[k] = temp
+            k = k - 1
+            cnt = cnt - 1
+        cnt = cnt +1
+    total = k+1
     for k in range(classes):
         for i in range(total):
-            s[i] = s[i]._replace(class_var=k)
-        s = sorted(s, key=lambda x: x.probs[x.index_var][x.class_var], reverse=True)
+            dets[i]['sort_class'] = k
+        dets[0:total] = sorted(dets[0:total],
+                               key=cmp_to_key(nms_comparator), reverse=False)
         for i in range(total):
-            if probs[s[i].index_var][k] == 0:
+            if dets[i]['prob'][k] == 0:
                 continue
-            a = boxes[s[i].index_var]
+            a = dets[i]['bbox']
             for j in range(i+1, total):
-                b = boxes[s[j].index_var]
+                b = dets[j]['bbox']
                 if _box_iou(a, b) > thresh:
-                    probs[s[j].index_var][k] = 0
-    return boxes, probs
+                    dets[j]['prob'][k] = 0
```

Comment:
How about using itertools.combinations?
```
for det_i, det_j in combinations(dets[0:total], 2):
    if det_i['prob'][k] == 0:
        ...
```

Reply:
It will check for ith det's kth prob mutliple times (for all combination of pair having ith det). Where as in previous implementation it will check once and skip all combinations of pair having ith det, if it is zero.